uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget-absence-types-amount-taken.component.js

```diff
@@ -0,0 +1,122 @@
+/* eslint-env amd */
+
+define([
+  'common/lodash',
+  'common/moment',
+  'leave-absences/shared/modules/components',
+  'leave-absences/shared/components/leave-requests-heatmap.component',
+  'leave-absences/shared/models/leave-request.model'
+], function (_, moment, components) {
+  components.component('leaveWidgetAbsenceTypesAmountTaken', {
+    bindings: {
+      title: '@',
+      absenceTypes: '<',
+      contactId: '<',
+      absencePeriod: '<',
+      leaveRequestStatuses: '<'
+    },
+    controller: absenceTypesTakenController,
+    controllerAs: 'absenceTypesTaken',
+    templateUrl: ['shared-settings', function (sharedSettings) {
+      return sharedSettings.sharedPathTpl + 'components/leave-widget/leave-widget-absence-types-amount-taken.html';
+    }]
+  });
+
+  absenceTypesTakenController.$include = ['$scope', 'LeaveRequest',
+    'OptionGroup'];
+
+  function absenceTypesTakenController ($scope, LeaveRequest, OptionGroup) {
+    var vm = this;
+
+    vm.leaveRequests = [];
+
+    vm.$onChanges = $onChanges;
+
+    /**
+     * Initializes the controller by emiting a child is loading event.
+     */
+    (function init () {
+      $scope.$emit('LeaveWidget::childIsLoading');
+    })();
+
+    /**
+     * Implements the $onChanges method for angular controllers. When bindings
+     * are ready for use, it loads all dependencies.
+     */
+    function $onChanges () {
+      if (areBindingsReady()) {
+        loadDependencies();
+      }
+    }
+
+    /**
+     * Returns true when the bindings are ready to be used.
+     *
+     * @return {Boolean}
+     */
+    function areBindingsReady () {
+      return vm.absenceTypes && vm.contactId && vm.absencePeriod &&
+        vm.leaveRequestStatuses && vm.leaveRequestStatuses.length;
+    }
+
+    /**
+     * Loads the status ids for absence types and leave requests for the contact.
+     * After loading the dependencies, it emits a child is ready event.
+     *
+     * @return {Promise}
+     */
+    function loadDependencies () {
+      return loadLeaveRequests()
+        .finally(function () {
```

Comment:
why `finally`? you want to emit the event even if there was an error?

Reply:
yes, otherwise we risk blocking the whole leave widget. If one of the components doesn't fire the child is ready event, the leave widget will keep waiting for it forever.